app/assets/javascripts/admin/addon/components/emoji-value-list.js

```diff
@@ -0,0 +1,186 @@
+import Component from "@ember/component";
+import I18n from "I18n";
+import { isEmpty } from "@ember/utils";
+import { on } from "discourse-common/utils/decorators";
+import { emojiUrlFor } from "discourse/lib/text";
+import { action, set } from "@ember/object";
+import { later, schedule } from "@ember/runloop";
+
+export default Component.extend({
+  classNameBindings: [":value-list"],
+  collection: null,
+  values: null,
+  validationMessage: null,
+  emojiPickerIsActive: false,
+  isEditorFocused: false,
+  emojiName: null,
+
+  init() {
+    this._super(...arguments);
+    this.set("collection", []);
+  },
+
+  @action
+  emojiSelected(code) {
+    this.set("emojiName", code);
+    this.set("emojiPickerIsActive", !this.emojiPickerIsActive);
+    this.set("isEditorFocused", !this.isEditorFocused);
+  },
+
+  @action
+  openEmojiPicker() {
+    this.set("isEditorFocused", !this.isEditorFocused);
+    later(() => {
+      this.set("emojiPickerIsActive", !this.emojiPickerIsActive);
+    }, 100);
+  },
+
+  @action
+  clearInput() {
+    this.set("emojiName", "");
+  },
+
+  @on("didReceiveAttrs")
+  _setupCollection() {
+    this.set("collection", this._splitValues(this.values));
+  },
+
+  _splitValues(values) {
+    if (values && values.length) {
+      const res = [];
+      const emojis = values.split("|");
+      emojis.forEach((emoji, index) => {
+        const object = {};
+        object.value = emoji;
+
+        object.emojiUrl = emojiUrlFor(emoji);
+        object.isEditable = true;
+
+        object.isEditing = false;
+
+        object.isLast = emojis.length - 1 === index;
+
+        res.push(object);
+      });
+
+      return res;
+    } else {
+      return [];
+    }
+  },
+
+  @action
+  editValue(index) {
+    const item = this.collection[index];
+    if (item.isEditable) {
+      set(item, "isEditing", !item.isEditing);
+      schedule("afterRender", () => {
+        const textbox = document.querySelector(
+          `[data-index="${index}"] .value-input`
+        );
+        if (textbox) {
+          textbox.focus();
+        }
+      });
+    }
+  },
+
+  @action
+  changeValue(index, newValue) {
+    const item = this.collection[index];
+
+    if (this._checkInvalidInput(newValue)) {
+      const oldValues = this.values.split("|");
```

Comment:
what if values is empty? can it be empty ?

Reply:
values are the site-setting value, if it is empty then `collection` will be empty. and if the `collection` is empty then there's no editing of the textbox

Reply:
so this line will never be called if this.values is empty ?

Reply:
yes thats true

Reply:
ok cool then